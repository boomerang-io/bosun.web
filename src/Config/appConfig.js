import { BASE_LAUNCH_ENV_URL } from "./platformUrlConfig";
// Look for the data injected into the HTML file from the Express app
// See server/app.js for implementation
export const APP_ROOT = window?._SERVER_DATA?.APP_ROOT ?? "/bosun";

export const PRODUCT_STANDALONE = process.env.PRODUCT_STANDALONE === "true" ?? false;;

// TODO
export const ROUTES = {};
export const appPath = {
  root: APP_ROOT,
  createPolicy: "/teams/:teamId/policy/create",
  createTemplate: "/templates/create",
  editPolicy: "/teams/:teamId/policy/edit/:policyId",
  editTemplate: "/templates/edit/:templateId",
  overview: "/teams/:teamId",
  templates: "/templates",
};

export const appLink = {
  policyTemplates: () => `/templates`,
  createPolicy: ({ teamId }) => `/teams/${teamId}/policy/create`,
  createTemplate: () => `/templates/create`,
  editPolicy: ({ teamId, policyId }) => `/teams/${teamId}/policy/edit/${policyId}`,
  editTemplate: ({ templateId }) => `/templates/edit/${templateId}`,
  teamOverview: ({ teamId }) => `/teams/${teamId}`,

  // CICD
  home: ({ activeTeamId }) => `${BASE_LAUNCH_ENV_URL}/cicd/apps/teams/${activeTeamId}`,
  components: ({ activeTeamId }) => `${BASE_LAUNCH_ENV_URL}/cicd/apps/teams/${activeTeamId}/components`,
  pipelines: ({ activeTeamId }) => `${BASE_LAUNCH_ENV_URL}/cicd/apps/teams/${activeTeamId}/pipelines`,
  scorecard: ({ activeTeamId }) => `${BASE_LAUNCH_ENV_URL}/cicd/apps/teams/${activeTeamId}/scorecard`,
  insights: ({ activeTeamId }) => `${BASE_LAUNCH_ENV_URL}/cicd/apps/teams/${activeTeamId}/insights`,
  componentModes: () => `${BASE_LAUNCH_ENV_URL}/cicd/apps/admin/component-modes`,
  properties: () => `${BASE_LAUNCH_ENV_URL}/cicd/apps/admin/properties`,
  scmRepositories: () => `${BASE_LAUNCH_ENV_URL}/cicd/apps/admin/repositories`,
  settings: () => `${BASE_LAUNCH_ENV_URL}/cicd/apps/admin/settings`,
  teamProperties: ({ activeTeamId }) => `${BASE_LAUNCH_ENV_URL}/cicd/apps/teams/${activeTeamId}/properties`,
  teamConfiguration: ({ activeTeamId }) => `${BASE_LAUNCH_ENV_URL}/cicd/apps/teams/${activeTeamId}/configuration`,

  // Lib
  lib: () => `${BASE_LAUNCH_ENV_URL}/cicd/apps/lib`,
};

export const FeatureFlag = {
  Standalone: "standalone",
};
