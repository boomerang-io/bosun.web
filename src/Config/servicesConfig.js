import axios, { CancelToken } from "axios";
import HTTPMethods from "Constants/httpMethods";
import portForwardMap from "../setupPortForwarding";

//const REACT_APP_PORT_FORWARD = process.env.REACT_APP_PORT_FORWARD;

export const BASE_SERVICE_ENV_URL =
  process.env.NODE_ENV === "production" ? window._SERVER_DATA && window._SERVER_DATA.BASE_SERVICE_ENV_URL : "/api";

export const PRODUCT_SERVICE_ENV_URL =
  process.env.NODE_ENV === "production" ? window._SERVER_DATA && window._SERVER_DATA.PRODUCT_SERVICE_ENV_URL : "/api";

export const CICD_PRODUCT_SERVICE_ENV_URL =
  process.env.NODE_ENV === "production"
    ? window._SERVER_DATA && window._SERVER_DATA.CICD_PRODUCT_SERVICE_ENV_URL
    : "/api";

const REACT_APP_PORT_FORWARD = process.env.REACT_APP_PORT_FORWARD;

/**
 * if port forwarding is enabled, then check to see if service is in config map
 * If it is, set the url request to be only the serviceContextPath so the url is relativet to the root of the app
 * CRA will proxy the request as seen in setupProxy.js
 * @param {string} baseUrl - base of the serivce url
 * @param {sring} serviceContextPath - additional path for the service context e.g. /admin
 */
function determineUrl(baseUrl, serviceContextPath) {
  if (REACT_APP_PORT_FORWARD && portForwardMap[serviceContextPath]) {
    return serviceContextPath;
  } else {
    return baseUrl + serviceContextPath;
  }
}
export const BASE_SERVICE_USERS_URL = determineUrl(BASE_SERVICE_ENV_URL, "/users");
export const BASE_SERVICE_PRODUCT_URL = determineUrl(PRODUCT_SERVICE_ENV_URL, "/policy");
export const BASE_CICD_PRODUCT_URL = determineUrl(CICD_PRODUCT_SERVICE_ENV_URL, "/ci");
// Product
export const SERVICE_REQUEST_STATUSES = {
  FAILURE: "failure",
  SUCCESS: "success",
};

export const serviceUrl = {
  deletePolicy: ({ policyId }) => `${BASE_SERVICE_PRODUCT_URL}/policies/${policyId}`,
  getCicdNavigation: ({ teamId }) => `${BASE_CICD_PRODUCT_URL}/navigation${teamId ? "?teamId=" + teamId : ""}`,
  getInsights: ({ teamId }) => `${BASE_SERVICE_PRODUCT_URL}/policies/insights?teamId=${teamId}`,
  getInsightsOverview: () => `${BASE_SERVICE_PRODUCT_URL}/policies/insights`,
  getPlatformNavigation: () => `${BASE_SERVICE_USERS_URL}/navigation`,
  getPolicy: ({ policyId }) => `${BASE_SERVICE_PRODUCT_URL}/policies/${policyId}`,
  getPolicyOverview: () => `${BASE_SERVICE_PRODUCT_URL}/policies`,
  getTeams: () => `${BASE_SERVICE_PRODUCT_URL}/teams`,
  getTeamPolicies: ({ teamId }) => `${BASE_SERVICE_PRODUCT_URL}/policies?teamId=${teamId}`,
  getTemplates: () => `${BASE_SERVICE_PRODUCT_URL}/templates`,
  getUserProfile: () => `${BASE_SERVICE_USERS_URL}/profile`,
  getValidateInfo: ({ policyId }) => `${BASE_SERVICE_PRODUCT_URL}/validate/info/${policyId}`,
  getViolations: ({ teamId }) => `${BASE_SERVICE_PRODUCT_URL}/policies/violations?teamId=${teamId}`,
  getViolationsOverview: () => `${BASE_SERVICE_PRODUCT_URL}/policies/violations`,
  patchUpdatePolicy: ({ policyId }) => `${BASE_SERVICE_PRODUCT_URL}/policies/${policyId}`,
  patchUpdatePolicyTemplate: ({ templateId }) => `${BASE_SERVICE_PRODUCT_URL}/templates/${templateId}`,
  postCreatePolicy: () => `${BASE_SERVICE_PRODUCT_URL}/policies`,
  postCreatePolicyTemplate: () => `${BASE_SERVICE_PRODUCT_URL}/templates`,
  postCreateTeam: () => `${BASE_SERVICE_PRODUCT_URL}/teams`,
};

export const cancellableResolver = ({ url, method, body, ...config }) => {
  // Create a new CancelToken source for this request
  const source = CancelToken.source();
  const promise = axios({ url, method, body, cancelToken: source.token, ...config });
  return { promise, cancel: () => source.cancel("cancel") };
};

export const resolver = {
  query: (url) => () => axios.get(url).then((response) => response.data),
  postMutation: (request) => axios.post(request),
  patchMutation: (request) => axios.patch(request),
  putMutation: (request) => axios.put(request),
  deletePolicy: ({ policyId }) =>
    cancellableResolver({ url: serviceUrl.deletePolicy({ policyId }), method: HTTPMethods.Delete }),
  patchUpdatePolicy: ({ policyId, body }) =>
    cancellableResolver({ url: serviceUrl.patchUpdatePolicy({ policyId }), method: HTTPMethods.Patch, data: body }),
  patchUpdatePolicyTemplate: ({ templateId, body }) =>
    cancellableResolver({
      url: serviceUrl.patchUpdatePolicyTemplate({ templateId }),
      method: HTTPMethods.Patch,
      data: body,
    }),
  postCreatePolicy: ({ body }) =>
    cancellableResolver({ url: serviceUrl.postCreatePolicy(), method: HTTPMethods.Post, data: body }),
  postCreatePolicyTemplate: ({ body }) =>
    cancellableResolver({ url: serviceUrl.postCreatePolicyTemplate(), method: HTTPMethods.Post, data: body }),
  postCreateTeam: ({ body }) =>
    cancellableResolver({ url: serviceUrl.postCreateTeam(), data: body, method: HTTPMethods.Post }),
};
