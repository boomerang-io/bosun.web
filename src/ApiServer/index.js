import { Server, Serializer, Model } from "miragejs";
import { inflections } from "inflected";
import uuid from "uuid/v4";
import { serviceUrl } from "Config/servicesConfig";
import * as fixtures from "./fixtures";

export function startApiServer({ environment = "test", timing = 0 } = {}) {
  inflections("en", function (inflect) {
    // Prevent pluralization bc our apis are weird
    inflect.irregular("policy", "policy");
    inflect.irregular("violation", "violation");
    inflect.irregular("template", "template");
    inflect.irregular("insights", "insights");
  });

  return new Server({
    environment,
    // Load in mock data
    fixtures,
    // Return the data as is, don't add a root key
    serializers: {
      application: Serializer.extend({
        root: false,
        embed: true,
      }),
    },
    // Register the data as a model so we can use the schema
    models: {
      templates: Model,
      policies: Model,
    },

    routes() {
      // Control how long the responses take to resolve
      this.timing = timing;

      // Allow unhandled requests on the current domain to pass through
      this.passthrough();

      this.get("/info", () => []);

      /**
       * Simple GET of static data
       */
      this.get(serviceUrl.getUserProfile(), (schema) => {
        return schema.db.profile[0];
      });
      this.get(serviceUrl.getCicdNavigation({}), (schema, request) => {
        return schema.db.cicdNavigation.map((item) => {
          if (typeof item.link === "string") {
            item.link = item.link.replace(":teamId", request.queryParams.teamId);
          }

          return item;
        });
      });

      this.get(serviceUrl.getPlatformNavigation(), (schema) => {
        return schema.db.navigation[0];
      });

      this.get(serviceUrl.getTeams(), (schema) => {
        return schema.db.teams;
      });

      this.get(serviceUrl.getPolicy({ policyId: ":policyId" }), (schema, request) => {
        let { policyId } = request.params;
        let activePolicy = schema.policies.find(policyId);
        return activePolicy;
      });

      this.get(serviceUrl.getValidateInfo({ policyId: ":policyId" }), (schema) => {
        return schema.db.validateInfo;
      });

      this.get(serviceUrl.getTemplates(), (schema) => {
        return schema.db.templates;
      });

      this.get(serviceUrl.getPolicyOverview(), (schema, request) => {
        let { teamId } = request.queryParams;
        return schema.db.policies.filter((policy) => policy.teamId === teamId);
      });

      this.get(serviceUrl.getViolationsOverview(), (schema) => {
        return schema.db.violations;
      });

      this.get(serviceUrl.getInsightsOverview(), (schema) => {
        return schema.db.insights;
      });

      this.post(serviceUrl.postCreatePolicy(), (schema, request) => {
        let body = JSON.parse(request.requestBody);
        return schema.policies.create({ ...body, id: uuid() });
      });

      this.post(serviceUrl.postCreatePolicyTemplate(), (schema, request) => {
        let body = JSON.parse(request.requestBody);
        return schema.templates.create({ ...body, id: uuid() });
      });

      this.patch(serviceUrl.patchUpdatePolicy({ policyId: ":policyId" }), (schema, request) => {
        let { policyId } = request.params;
        let body = JSON.parse(request.requestBody);
        let activePolicy = schema.policies.find(policyId);
        activePolicy.update(body);
        return schema.policies.all();
      });

      this.patch(serviceUrl.patchUpdatePolicyTemplate({ templateId: ":templateId" }), (schema, request) => {
        let { templateId } = request.params;
        let body = JSON.parse(request.requestBody);
        let activeTemplate = schema.templates.find(templateId);
        activeTemplate.update(body);
        return schema.templates.all();
      });

      this.delete(serviceUrl.deletePolicy({ policyId: ":policyId" }), (schema, request) => {
        let { policyId } = request.params;
        let activePolicy = schema.policies.find(policyId);
        activePolicy.destroy({ id: policyId });
        return schema.policies.all();
      });
    },
  });
}
