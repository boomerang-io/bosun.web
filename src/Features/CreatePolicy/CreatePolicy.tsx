import React, { useState } from "react";
import { useAppContext } from "Hooks";
import { useQuery, useMutation, queryCache } from "react-query";
import { Helmet } from "react-helmet";
import {
  notify,
  ToastNotification,
  ErrorMessage,
  TextInputSkeleton,
} from "@boomerang-io/carbon-addons-boomerang-react";
import CreateEditPolicyForm from "Components/CreateEditPolicyForm";
import CreateEditPolicyHeader from "Components/CreateEditPolicyHeader";
import DefinitionSkeleton from "Components/DefinitionSkeleton";
import { serviceUrl, resolver } from "Config/servicesConfig";
import { appLink } from "Config/appConfig";
import { POLICY_INTERACTION_TYPES } from "Constants";
import { PolicyDefinition, CreatePolicyData, PolicyDefinitionTemplate } from "Types";
import styles from "./createPolicy.module.scss";

type Props = {
  history: any;
  match: any;
};

export function CreatePolicy({ history, match }: Props) {
  const { activeTeam } = useAppContext();

  const definitionsUrl = serviceUrl.getTemplates();
  const { params } = match;

  const [errors, setErrors] = useState<any>({});
  const [inputs, setInputs] = useState<any>({});
  const [name, setName] = useState("");
  const cancelRequestRef = React.useRef<any>();

  const { data: definitionsData, isLoading, error } = useQuery<Array<PolicyDefinitionTemplate>, any>({
    queryKey: definitionsUrl,
    queryFn: resolver.query(definitionsUrl),
  });

  const [createPolicyMutation, { isLoading: createIsLoading }] = useMutation(
    (args: { body: CreatePolicyData }) => {
      const { promise, cancel } = resolver.postCreatePolicy(args);
      cancelRequestRef.current = cancel;
      return promise;
    },
    {
      onSuccess: () => queryCache.invalidateQueries(definitionsUrl),
    }
  );
  const createPolicy = async () => {
    let policyObject: CreatePolicyData = {
      name: name,
      teamId: activeTeam?.id ?? "",
      definitions: [],
    };

    definitionsData?.forEach((definition: PolicyDefinitionTemplate) => {
      let newDefinition: PolicyDefinition = {
        policyTemplateId: definition.id,
        rules: [],
      };
      let rules = [];
      const definitionRows = inputs[definition.key];
      for (let row in definitionRows) {
        rules.push(definitionRows[row]);
      }
      newDefinition["rules"] = rules;
      policyObject.definitions.push(newDefinition);
    });

    try {
      await createPolicyMutation({ body: policyObject });
      notify(<ToastNotification kind="success" title="Policy Created" subtitle="Policy was successfully created" />);
      navigateBack();
    } catch (e) {
      notify(<ToastNotification kind="error" title="Something's Wrong" subtitle="Request to create policy failed" />);
    }
  };

  // State updates

  const setError = (error: any) => {
    setErrors((prevState: any) => ({ ...prevState, ...error }));
  };

  const setInput = async ({ event: e, definitionKey, uuid }: any) => {
    const { name, value } = e.target;
    await setInputs((prevState: any) => {
      const prevStateDefinitionRows = prevState[definitionKey] ? prevState[definitionKey][uuid] : {};
      return {
        ...prevState,
        [definitionKey]: {
          ...prevState[definitionKey],
          [uuid]: { ...prevStateDefinitionRows, [name]: value },
        },
      };
    });
    validateRow(definitionKey);
  };

  const removeRow = async ({ definitionKey, uuid }: any) => {
    let definitionRows = { ...inputs[definitionKey] };
    if (definitionRows) {
      delete definitionRows[uuid];
      await setInputs((prevState: any) => ({
        ...prevState,
        [definitionKey]: definitionRows,
      }));
      validateRow(definitionKey);
    }
  };

  const validateRow = (definitionKey: any) => {
    const definitionRows = inputs[definitionKey] || {};
    const definitionRowsInputCount = Object.keys(definitionRows).reduce((accum, uuid) => {
      const inputCount = Object.values(definitionRows[uuid]).filter(Boolean).length;
      accum += inputCount;
      return accum;
    }, 0);

    // Each row should have the same number of inputs as the number of inputs in the policy definition rules
    const matchingDefintion = definitionsData?.find((definition: any) => definition.key === definitionKey);
    const isInvalid =
      matchingDefintion &&
      Object.keys(definitionRows).length * matchingDefintion.rules.length !== definitionRowsInputCount;
    setErrors((prevState: any) => ({ ...prevState, [definitionKey]: isInvalid }));
  };

  // Local methods

  const navigateBack = () => {
    history.push(appLink.teamOverview({ teamId: params.teamId }));
  };

  const form = {
    name,
    inputs,
    errors,
    setName,
    setInput: setInput,
    setError: setError,
    removeRow,
    validateRow,
    onCancel: cancelRequestRef.current,
    affirmativeAction: createPolicy,
    isPerformingAffirmativeAction: createIsLoading,
  };

  return (
    <div className={styles.container}>
      <Helmet>
        <title>Create Policy - Bosun Policies</title>
      </Helmet>
      <CreateEditPolicyHeader
        form={form}
        navigateBack={navigateBack}
        type={POLICY_INTERACTION_TYPES.CREATE}
        hasError={Boolean(error)}
      />
      {isLoading ? (
        <div className={styles.skeletonsContainer}>
          <TextInputSkeleton className={styles.textInputSkeleton} />
          <DefinitionSkeleton />
          <DefinitionSkeleton />
          <DefinitionSkeleton />
        </div>
      ) : error ? (
        <div style={{ marginTop: "2rem" }}>
          <ErrorMessage />
        </div>
      ) : (
        <CreateEditPolicyForm form={form} definitions={definitionsData ?? []} />
      )}
    </div>
  );
}

export default CreatePolicy;
