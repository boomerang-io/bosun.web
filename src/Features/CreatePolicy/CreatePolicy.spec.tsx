import React from "react";
import { Response } from "miragejs";
import { fireEvent } from "@testing-library/react";
import { queryCaches } from "react-query";
import { startApiServer } from "ApiServer";
import CreatePolicy from "../CreatePolicy";
import { serviceUrl } from "Config/servicesConfig";
import { appLink } from "Config/appConfig";

const route = appLink.createPolicy({ teamId: "5a8b331e262a70306622df73" });
const props = {
  match: { params: { policyId: "5cd49adff6ea74a9bb6adef3" } },
  history: {},
};
let server: any;

beforeEach(() => {
  server = startApiServer();
});

afterEach(() => {
  server.shutdown();
  queryCaches.forEach((queryCache) => queryCache.clear());
});

describe("CreatePolicy --- Snapshot", () => {
  it("renders correctly", async () => {
    const { baseElement, findByText } = rtlRouterRender(<CreatePolicy {...props} />, { route });
    await findByText(/Create Policy/i);
    await findByText(/Static Code Analysis/i);
    expect(baseElement).toMatchSnapshot();
  });
});

describe("CreatePolicy --- RTL", () => {
  test("renders error message when fetching definitions failed", async () => {
    server.get(serviceUrl.getTemplates(), () => {
      return new Response(404, {}, { data: { status: 404 } });
    });
    const { findByText } = rtlRouterRender(<CreatePolicy {...props} />, { route });
    const errorMessage = await findByText("Oops, something went wrong.");
    expect(errorMessage).toBeInTheDocument();
  });

  test("it is able to create policy only after adding a name", async () => {
    const { getByPlaceholderText, findByTestId, findByPlaceholderText } = rtlRouterRender(<CreatePolicy {...props} />, {
      route,
    });

    const createButton = await findByTestId("policy-header-affirmative-action");

    expect(createButton).toBeDisabled();
    await findByPlaceholderText(/name/i);
    const nameInput = getByPlaceholderText(/name/i);
    fireEvent.change(nameInput, { target: { value: "test" } });

    expect(createButton).toBeEnabled();
  });

  test("create button is disabled while creating", async () => {
    const { getByPlaceholderText, findByTestId, findByText, findByPlaceholderText } = rtlContextRouterRender(
      <CreatePolicy {...props} />,
      {
        route,
      }
    );

    const createButton = await findByTestId("policy-header-affirmative-action");
    await findByPlaceholderText(/name/i);
    const nameInput = getByPlaceholderText(/name/i);
    fireEvent.change(nameInput, { target: { value: "test" } });
    expect(createButton).toBeEnabled();

    fireEvent.click(createButton);
    expect(await findByText(/Create Policy/)).toBeInTheDocument();
  });
});
