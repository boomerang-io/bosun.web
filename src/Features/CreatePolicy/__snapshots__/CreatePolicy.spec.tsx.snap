// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreatePolicy --- Snapshot renders correctly 1`] = `
<body
  id="app"
>
  <div>
    <div
      class="container"
    >
      <header
        class="bx--bmrg-feature-header"
        style="top: 3rem;"
      >
        <section
          class="bx--bmrg-feature-header__content"
        >
          <div
            class="bx--bmrg-feature-header__header"
          >
            <h1
              class="bx--bmrg-feature-header-title"
            >
              Create Policy
            </h1>
            
          </div>
        </section>
        <section
          class="buttons"
        >
          <button
            class="button bx--btn bx--btn--field bx--btn--secondary"
            tabindex="0"
            type="button"
          >
            Cancel
          </button>
          <button
            class="button bx--btn bx--btn--field bx--btn--primary bx--btn--disabled"
            data-testid="policy-header-affirmative-action"
            disabled=""
            tabindex="0"
            type="submit"
          >
            Create
            <svg
              aria-hidden="true"
              aria-label="Create"
              class="bx--btn__icon"
              fill="currentColor"
              focusable="false"
              height="16"
              preserveAspectRatio="xMidYMid meet"
              role="img"
              viewBox="0 0 32 32"
              width="16"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M17 15L17 8 15 8 15 15 8 15 8 17 15 17 15 24 17 24 17 17 24 17 24 15z"
              />
            </svg>
          </button>
        </section>
      </header>
      <div
        class="container"
      >
        <div
          class="content"
        >
          <div
            class="bx--bmrg-text-input"
          >
            <div
              class="bx--form-item bx--text-input-wrapper"
            >
              <label
                class="bx--label"
                for="name"
              >
                <div
                  class="bx--bmrg-text-input__label"
                >
                  <div>
                    Name
                  </div>
                </div>
              </label>
              <div
                class="bx--text-input__field-outer-wrapper"
              >
                <div
                  class="bx--text-input__field-wrapper"
                >
                  <input
                    class="bx--text-input bx--text__input"
                    id="name"
                    name="name"
                    placeholder="Name"
                    title="Name"
                    type="text"
                    value=""
                  />
                </div>
              </div>
            </div>
          </div>
          <section
            class="section"
          >
            <h2>
              Static Code Analysis
            </h2>
            <p>
              The following policy metrics are validated from SonarQube data collected in your CI pipeline.
            </p>
            <div
              class="rowsContainer"
            />
            <button
              class="bx--btn bx--btn--field bx--btn--ghost"
              style="margin-left: -0.625rem;"
              tabindex="0"
              type="button"
            >
              Add Rule
              <svg
                aria-hidden="true"
                aria-label="Add Rule"
                class="bx--btn__icon"
                fill="currentColor"
                focusable="false"
                height="16"
                preserveAspectRatio="xMidYMid meet"
                role="img"
                viewBox="0 0 32 32"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M17 15L17 8 15 8 15 15 8 15 8 17 15 17 15 24 17 24 17 17 24 17 24 15z"
                />
              </svg>
            </button>
          </section>
          <section
            class="section"
          >
            <h2>
              asdadsf
            </h2>
            <p>
              asdsdaf
            </p>
            <div
              class="rowsContainer"
            />
            <button
              class="bx--btn bx--btn--field bx--btn--ghost"
              style="margin-left: -0.625rem;"
              tabindex="0"
              type="button"
            >
              Add Rule
              <svg
                aria-hidden="true"
                aria-label="Add Rule"
                class="bx--btn__icon"
                fill="currentColor"
                focusable="false"
                height="16"
                preserveAspectRatio="xMidYMid meet"
                role="img"
                viewBox="0 0 32 32"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M17 15L17 8 15 8 15 15 8 15 8 17 15 17 15 24 17 24 17 17 24 17 24 15z"
                />
              </svg>
            </button>
          </section>
          <section
            class="section"
          >
            <h2>
              Package Safe List
            </h2>
            <p>
              The following package and artifact validation is validated against scans by JFrog Xray. Artifact and Version are validated through regular expression.
            </p>
            <div
              class="rowsContainer"
            />
            <button
              class="bx--btn bx--btn--field bx--btn--ghost"
              style="margin-left: -0.625rem;"
              tabindex="0"
              type="button"
            >
              Add Rule
              <svg
                aria-hidden="true"
                aria-label="Add Rule"
                class="bx--btn__icon"
                fill="currentColor"
                focusable="false"
                height="16"
                preserveAspectRatio="xMidYMid meet"
                role="img"
                viewBox="0 0 32 32"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M17 15L17 8 15 8 15 15 8 15 8 17 15 17 15 24 17 24 17 17 24 17 24 15z"
                />
              </svg>
            </button>
          </section>
          <section
            class="section"
          >
            <h2>
              Security Issue Analysis
            </h2>
            <p>
              Create rules to guide the general health of your repositories with respect to security and other vulnerabilities.
            </p>
            <div
              class="rowsContainer"
            />
            <button
              class="bx--btn bx--btn--field bx--btn--ghost"
              style="margin-left: -0.625rem;"
              tabindex="0"
              type="button"
            >
              Add Rule
              <svg
                aria-hidden="true"
                aria-label="Add Rule"
                class="bx--btn__icon"
                fill="currentColor"
                focusable="false"
                height="16"
                preserveAspectRatio="xMidYMid meet"
                role="img"
                viewBox="0 0 32 32"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M17 15L17 8 15 8 15 15 8 15 8 17 15 17 15 24 17 24 17 17 24 17 24 15z"
                />
              </svg>
            </button>
          </section>
          <section
            class="section"
          >
            <h2>
              CVE Safe List
            </h2>
            <p>
              The Common Vulnerabilities & Exposures (CVE) are validated against JFrog Xray Security Issues. Artifact and Version are validated through regular expression.
            </p>
            <div
              class="rowsContainer"
            />
            <button
              class="bx--btn bx--btn--field bx--btn--ghost"
              style="margin-left: -0.625rem;"
              tabindex="0"
              type="button"
            >
              Add Rule
              <svg
                aria-hidden="true"
                aria-label="Add Rule"
                class="bx--btn__icon"
                fill="currentColor"
                focusable="false"
                height="16"
                preserveAspectRatio="xMidYMid meet"
                role="img"
                viewBox="0 0 32 32"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M17 15L17 8 15 8 15 15 8 15 8 17 15 17 15 24 17 24 17 17 24 17 24 15z"
                />
              </svg>
            </button>
          </section>
        </div>
        <div
          class="containerGraphics"
        >
          <svg
            class="graphic"
            data-name="Layer 1"
            id="prefix__Layer_1"
            viewBox="0 0 202.29 228.51"
          >
            <defs>
              <style>
                .prefix__cls-2{fill:#20d5d2}.prefix__cls-3{fill:none;stroke:#323c3c;stroke-linecap:round;stroke-miterlimit:10;stroke-width:5px}
              </style>
            </defs>
            <title>
              Artboard 54
            </title>
            <path
              d="M179.68 167.15a5.32 5.32 0 014 8.8l-24.07 27.79a5.34 5.34 0 01-4 1.83H22.31a5.32 5.32 0 01-3.41-9.41l33.1-27.5a5.28 5.28 0 013.4-1.23l65.52-.15c1.32 5.69 7.52 16.94 7.52 16.94s7-11.47 8.4-17z"
              fill="#f2f4f8"
            />
            <path
              class="prefix__cls-2"
              d="M98.2 44.67a27.15 27.15 0 114.12 24.71 27.15 27.15 0 01-4.12-24.71zm39.16 12.74a14 14 0 10-17.69 9 14.05 14.05 0 0017.69-9zM130.88 110.1l-.46 13.28-52.97-17.24 7.46-11 45.97 14.96zM129.25 157.41a8.21 8.21 0 017.8 8.37 7.5 7.5 0 01-.23 1.46H120.9a7.9 7.9 0 01-.24-1.88 8.17 8.17 0 018.38-8zM128.42 184.22s-6.2-11.25-7.52-16.94h15.92c-1.37 5.47-8.4 16.94-8.4 16.94zM46.91 135.77a8.21 8.21 0 0111.4-2.07l.33.25a8.19 8.19 0 011.74 11.17C57.8 148.83 43 155.75 43 155.75s1.32-16.26 3.91-19.98z"
            />
            <path
              class="prefix__cls-3"
              d="M120.9 167.28l-65.52.15a5.28 5.28 0 00-3.4 1.23l-33.08 27.5a5.32 5.32 0 003.41 9.41h133.31a5.34 5.34 0 004-1.83L183.71 176a5.32 5.32 0 00-4-8.8l-42.86.09M132.41 27.25A27.15 27.15 0 1198.2 44.67a27.15 27.15 0 0134.21-17.42zM137.03 13.07l-4.62 14.18"
            />
            <path
              class="prefix__cls-3"
              d="M102.32 69.38L84.91 95.14l-7.46 11-18.81 27.81M131.95 79.02l-1.07 31.08-.46 13.28-1.17 34.03M84.91 95.13l45.96 14.97M77.45 106.14l52.97 17.24M137.36 57.41a14 14 0 11-9-17.69 14.05 14.05 0 019 17.69zM46.91 135.77a8.21 8.21 0 0111.4-2.07l.33.25a8.19 8.19 0 011.74 11.17C57.8 148.83 43 155.75 43 155.75s1.32-16.26 3.91-19.98z"
            />
            <path
              class="prefix__cls-3"
              d="M129.25 157.41H129a8.17 8.17 0 00-8.38 8 7.9 7.9 0 00.24 1.88c1.32 5.69 7.52 16.94 7.52 16.94s7-11.47 8.4-17a7.5 7.5 0 00.23-1.46 8.21 8.21 0 00-7.76-8.36z"
            />
          </svg>
        </div>
      </div>
    </div>
  </div>
</body>
`;
